Prevent style rules loss by not moving already-hydrated <style> tags

gregjacobs opened the pull request:
<!--
Thanks for your interest in the project. I appreciate bugs filed and PRs submitted!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: This PR fixes emotion-js/emotion#2210. The issue is that `<style>` elements are moved in the DOM, and that causes them to lose their `sheet.cssRules` array in production ("speedy") mode

<!-- Why are these changes necessary? -->

**Why**: Multiple micro-frontends can use the same Emotion key, which causes Emotion to accidentally clobber the first micro-frontend's styles by loading the second's

<!-- How were these changes implemented? -->

**How**: By simply detecting when we have already processed an existing `<style>` tag within a `StyleSheet` instance, we know never to move that particular `<style>` tag in the DOM again.

Emotion was previously moving the `<style>` tag to the same location in the DOM from the `StyleSheet._insertTag` method, and this caused the loss of styles.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation (N/A)
- [X] Tests
- [X] Code complete
- [X] Changeset <!-- This is necessary if your changes should release any packages. Run `yarn changeset` to create a changeset -->

<!-- feel free to add additional comments -->

Andarist commented:
This solution has a problem - each sheet associated with the same key will "own" style elements and if you flush a sheet the styles will be gone but the cache owning another sheet won't be notified by that and it won't be able to reinsert some styles.

Why can't you use different keys **or** reuse the same instance across your page?

gregjacobs commented:
@Andarist So I was thinking about the flush() issue too after I went to bed last night and was going to get up this morning and change the implementation, haha

I think the original solution I proposed in https://github.com/emotion-js/emotion/issues/2210#issuecomment-766427504 is best then, where basically only one `StyleSheet` instance will ever own a given `<style>` tag, and `flush()` will work correctly. 

Essentially this will make it so that the first Emotion instance for a given `key` will own any existing SSR `<style>` tag (if there is one), and subsequent instances will just create their own `<style>` tags from scratch. The subsequent instances will potentially duplicate some of the first `<style>` tag's styles, but I think that's okay - the page will still render correctly, and as you said, sharing the same key isn't really a "supported" use case anyway so I think a couple of duplicated style rules is okay. This will just make micro-frontend apps which happen to share the same key render correctly.

----

To answer your question about why we can't use different keys or the same Emotion instance on our page, I actually had written a small comment in https://github.com/emotion-js/emotion/issues/2210#issuecomment-766428070 of why it's important for micro-frontends for Emotion to support this use case in general, but to give a little more color to it, we actually don't own the micro-frontends and there will be many of them published by many different teams. I work for a large company, and in one division alone there are 1200 developers. Imagine trying to get them all to coordinate and stay in sync on Emotion keys, or even implementing a pattern to prevent them from clashing in each of their apps 😅  

The best solutions for us are most-often technical ones rather than human/pattern ones, which is why having Emotion do as-close-to-the-right-thing-as-possible (even if it's not a fully-supported use case) would be best.

gregjacobs commented:
Ok, so I added the new flag to make sure that all `<style>` tags are always owned by exactly one `StyleSheet` instance. I did this by adding a `data-h` attribute ("hydrated") to all `<style>` tags that have been processed. 

All tests pass, and I added multi-instance tests as well

Was just thinking though: could I possibly leverage the existing `data-s` attribute to determine when a `<style>` tag is already owned by a StyleSheet? What exactly does `data-s` represent? (If not, can leave it as `data-h`)

gregjacobs commented:
Btw, the changes in source code are just 2 lines. Had to update a bunch of snapshots though 😅

gregjacobs commented:
Hey @Andarist, any way to get that last CI check to pass? Seems to be stuck? 

Best,
Greg

Andarist commented:
This still has a problem with Emotion 11 hijacking some Emotion 10 styles as Emotion 10 doesn't put those data markers into the DOM at all. I'll try to figure out how to best handle this, considering all of those scenarios - unfortunately, I'm kinda burned out on Emotion a little bit right now so I'm not sure when I will be able to think this through

gregjacobs commented:
Believe this has already been fixed (discussion and PR link in https://github.com/emotion-js/emotion/issues/2210). Closing

Andarist commented:
@gregjacobs thank you for cleaning this up and for filing this PR in the first place. I'm sorry that we didn't get to your PR sooner so we could figure out the final solution and that the other (newer) PR has been pushed forward. It all really boiled down to your PR being created in a less convenient time for us (we had less free time to give to the project at that point in time) and this has slipped our attention.

gregjacobs commented:
Hey @Andarist. It's all good man, thanks for looking at it in the first place! I had lost track of this PR as well actually as I had worked around the issue on my side by ensuring a single global Emotion instance. Glad someone else was able to take the reins though and come up with a good solution! 

I appreciate the attention, kind words, and follow up, and I'll be back to contribute again in the future if the need should arise!

All the best,
Greg
